### PR TITLE
Wholesale refactor of the map source tests

### PR DESCRIPTION
**Pull request by nabobalis**
Moved all the tests for the sources to follow how they are laid out under the source folder aka per mission/satellite/observatory. 

The only thing that has changed from current main is that:

1. Added tests for Yohkoh
2. This has lead to two changes to that map source file. 

Otherwise everything else is just moving tests and some doc tweaks.

### CODE REVIEW

File: sunpy/map/sources/yohkoh.py [hunk 1]
```diff
@@ -57,23 +57,22 @@ def detector(self):
...
-        # the Sun from the spacecraft.  Note that the small angle approximation
+        # the Sun from the spacecraft. Note that the small angle approximation
         # is used, and the solar radius stored in SXT FITS files is in arcseconds.
-
         if 'solar_r' in self.meta:
             dsun = constants.radius / (np.deg2rad(self.meta['solar_r'] / 3600.0))
         else:
             dsun = constants.au
-        return self.meta.get('dsun_obs', dsun)
+        return self.meta.get('dsun_obs', dsun) * u.m
```
Comment (nabobalis) [inline sunpy/map/sources/yohkoh.py:75]: I had to do this.

File: sunpy/map/sources/yohkoh.py [hunk 2]
```diff
@@ -89,8 +88,15 @@ def wavelength(self):
...
     def is_datasource_for(cls, data, header, **kwargs):
-        """Determines if header corresponds to an SXT image"""
+        """
+        Determines if header corresponds to an SXT image.
+        """
         return header.get('instrume') == 'SXT'
+
+    @property
+    def date(self):
+        return Time(self.meta.get('DATE_OBS'))
```
Comment (nabobalis) [inline sunpy/map/sources/yohkoh.py:102]: I had to do this.

File: sunpy/map/sources/yohkoh.py (hunk 1 above)
Comment (ayshih) [reply sunpy/map/sources/yohkoh.py:75]: Er, neither is correct.  `self.meta.get('dsun_obs')` is a bare value (in meters), while `dsun` is already a `Quantity`.

**Comment by dstansby** [comment]
General thoughts on this PR:
- Can we get rid of the whitespace changes in the header files? They seem unnesseseary and make the PR harder to review
- Can you put any behaviour changes or bugfixes in individual PRs, leaving this on purely for refactoring? That way we can much more easily keep track of changelogs/backports etc.

**Comment by Cadair** [comment]
The whitespace changes in the header files might actually be bad (I haven't looked closely) but a real FITS header always has 80 char per card padded with whitespace. So I would rather keep them as close to "copied out of file" as possible.

File: sunpy/map/sources/yohkoh.py (hunk 1 above)
Comment (nabobalis) [reply sunpy/map/sources/yohkoh.py:75]: I will fix this.

**Comment by nabobalis** [comment]
> The whitespace changes in the header files might actually be bad (I haven't looked closely) but a real FITS header always has 80 char per card padded with whitespace. So I would rather keep them as close to "copied out of file" as possible.

I can undo the ones I changed but the ones I created don't have them.

**Comment by nabobalis** [comment]
> General thoughts on this PR:
> 
> * Can we get rid of the whitespace changes in the header files? They seem unnesseseary and make the PR harder to review

Will do

> * Can you put any behaviour changes or bugfixes in individual PRs, leaving this on purely for refactoring? That way we can much more easily keep track of changelogs/backports etc.

The only bugfix I can see is the yohkoh if that even counts.

**Comment by dstansby** [comment]
Yeah, it was the changes you pointed out in your own comments that I thought deserved separate PR(s?)

**Comment by nabobalis** [comment]
I would rather they stay here since they are one line changes.

File: sunpy/map/sources/yohkoh.py (hunk 2 above)
Comment (nabobalis) [reply sunpy/map/sources/yohkoh.py:102]: I suspect this is a bug in mapbase but this was quicker.

File: sunpy/map/sources/yohkoh.py (hunk 1 above)
Comment (nabobalis) [reply sunpy/map/sources/yohkoh.py]: What do you think about the comment above it? 

If that is true, we don't in fact want to return the meta keyword value at all?

**Comment by nabobalis** [comment]
I will come back for these header files in the future.

File: sunpy/map/sources/yohkoh.py [hunk 3]
```diff
@@ -57,23 +56,22 @@ def detector(self):
...
-        # the Sun from the spacecraft.  Note that the small angle approximation
+        # the Sun from the spacecraft. Note that the small angle approximation
         # is used, and the solar radius stored in SXT FITS files is in arcseconds.
-
         if 'solar_r' in self.meta:
             dsun = constants.radius / (np.deg2rad(self.meta['solar_r'] / 3600.0))
         else:
             dsun = constants.au
-        return self.meta.get('dsun_obs', dsun)
+        return dsun
```
Comment (nabobalis) [inline sunpy/map/sources/yohkoh.py:74]: To align with the comment, I decided to calculate the value and return that instead.

Comment (ayshih) [reply sunpy/map/sources/yohkoh.py:74]: Well...  The test header has a usable (and presumably) correct value for `DSUN_OBS`...

Comment (ayshih) [reply sunpy/map/sources/yohkoh.py:74]: I think we should get rid of this overridden `dsun` property entirely.  As currently implemented, `.dsun` is not used to construct the observer coordinate, so the change of inconsistency is high.  If `DSUN_OBS` is missing, we need `.dsun` to give the same answer as `.observer_coordinate.radius`.  If `DSUN_OBS` is present but wildly wrong, that error needs to be addressed during the creation of the observer coordinate, not in this property.

**Comment by ayshih** [comment]
Bleh, these values in the SXT header ought to be equal:
```
SOLAR_B0=              3.06603
HGLT_OBS=              0.00000
```
`HGLT_OBS` is clearly incorrect.  We'll need to override `._supported_observer_coordinates()`.

**Comment by ayshih** [comment]
Something like this:
```python
    @property
    def _supported_observer_coordinates(self):
        return [(('hgln_obs', 'solar_b0', 'dsun_obs'), {'lon': self.meta.get('hgln_obs'),
                                                        'lat': self.meta.get('solar_b0'),
                                                        'radius': self.meta.get('dsun_obs'),
                                                        'unit': (u.deg, u.deg, u.m),
                                                        'frame': "heliographic_stonyhurst"})]
            + super()._supported_observer_coordinates
```

**Comment by dstansby** [comment]
I'm going to request again that any bug fixes or behaviour changes go in individual PRs with appropriate names; the overheads trying to review a PR that contains refactoring and changes are large, make it less likely for reviewers to review code, and much easier for bugs to slip through the cracks during review.

**Comment by nabobalis** [comment]
The only code changes are in the Yohkoh map file. That I think is isolated enough to not need pulling out.

**Comment by nabobalis** [comment]
I think since we need to backport this and we didn't backport the test files changes, I am wondering if it's time to backport those now.

File: sunpy/map/sources/yohkoh.py (hunk 3 above)
Comment (nabobalis) [reply sunpy/map/sources/yohkoh.py]: This is all gone now and using your suggested fixes.

File: sunpy/map/mapbase.py [hunk 4]
```diff
@@ -786,10 +786,9 @@ def nickname(self, n):
         self._nickname = n
 
     def _get_date(self, key):
-        time = self.meta.get(key, None)
-        if time is None:
+        time = self.meta.get(key)
+        if not time:
```
Comment (nabobalis) [inline sunpy/map/mapbase.py:790]: This might be a bug?

File: sunpy/map/mapbase.py [hunk 5]
```diff
@@ -786,10 +786,9 @@ def nickname(self, n):
...
             return
-
```
Comment (dstansby) [inline sunpy/map/mapbase.py:792]: What's the context for this change?

**Review by dstansby — CHANGES_REQUESTED** [review]
I'm going to mark this as changes again because:
- We haven't agreed to apply black to the codebase, and there's lots of that in this PR. @nabobalis if you're intent on doing this, please can we have clearly labelled commit messages that *only* apply black and nothing else?
- I'm going to insist that any changes to actual code that aren't refactoring go in their own PRs for visibility and ease of review. In this case I think that's just the Yokoh changes

**Comment by dstansby** [comment]
Sorry, I should have said codestyle changes in general, not just black.

**Comment by nabobalis** [comment]
I can remove actual code changes to a specific PR but the rest of the PR is a code refactor, it is going to look different and there is no simple way to commit just that

**Comment by dstansby** [comment]
I'm not against codestyle changes in general, but it's really tricky to review when they're mixed in with other changes. So commits specifically for code style updates that have "Codestyle updates" or simliar in the commit message, and are *just* codestyle updates make it much easier to review stuff (and bisect any issues in the future).

File: sunpy/map/mapbase.py (hunk 5 above)
Comment (nabobalis) [reply sunpy/map/mapbase.py:792]: The key could be empty but not None.

**Comment by dstansby** [comment]
Just coming back to this, and I think it's actually better to have one test file per source, so it's easier to check that each map source is being tested in the same way. Long term it might be nicer to parametrize all the map source tests in one file. So I'd advocate for closing this without merging, and if there's small bits that aren't re-organising tests putting them in new PRs, thoughts?

**Comment by dstansby** [comment]
Sorry, I don't follow - do the same what for all map sources?

**Comment by nabobalis** [comment]
Either the map sources should be one class per file or the tests should be collected together to match the map sources files

**Comment by dstansby** [comment]
My preference would be single file per spacecraft, which would mean merging the test files. If we go this way, tests for each map class should live within their own test class.

**Comment by dstansby** [comment]
Worth closing this PR and redoing it in a new (or several new) PR if it's still relevant?

**Comment by nabobalis** [comment]
I will handle it.

**Pull request closed without merging**